### ISSUE TITLE
NVIDIA nv0 training stalls beyond 60s on quantized pool model with batch normalization and bfloat16

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#316]
Public `recipe.train().optimizer(adamw).lr(0.001).seed(17856690329093).epochs(1).log(all).stop(0.0).bf(16).run()` on `backend:nv0` exceeds the slow-runtime threshold (65s, stable). The earliest failure is `recipe.train().run()` stalling at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA `ptx`/`ptxas` branch (`recipe.rs` ~2868-2896) before any epoch arithmetic. The model (`recipe.model().pool(2).gelu().norm(batch).iq(2).xxs .pool(2).norm(batch).qi(4).k.m .loss(bce)`) uses quantized pool nodes with batch normalization and `.bf(16)`; the loader (`data/temporal/daily_shards` broadcast) is ignored. No existing nv0 issue covers this pure quantized-pool native-compilation path (271 = residual, 308 = pool/attention, 291 = pool/GRU, 314 = layer/pool). Expected sub-second epoch time for this small temporal load; 65s indicates the hang is native artifact compilation, not throughput.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `57cb73c0d45fb976`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling-small:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 training stalls beyond 60s on quantized pool model with batch normalization and bfloat16",
  "body": "Public `recipe.train().optimizer(adamw).lr(0.001).seed(17856690329093).epochs(1).log(all).stop(0.0).bf(16).run()` on `backend:nv0` exceeds the slow-runtime threshold (65s, stable). The earliest failure is `recipe.train().run()` stalling at `compile_model` -> `NativeModelIr::from_graph` -> `emit` -> `compile_native_artifact` on the NVIDIA `ptx`/`ptxas` branch (`recipe.rs` ~2868-2896) before any epoch arithmetic. The model (`recipe.model().pool(2).gelu().norm(batch).iq(2).xxs .pool(2).norm(batch).qi(4).k.m .loss(bce)`) uses quantized pool nodes with batch normalization and `.bf(16)`; the loader (`data/temporal/daily_shards` broadcast) is ignored. No existing nv0 issue covers this pure quantized-pool native-compilation path (271 = residual, 308 = pool/attention, 291 = pool/GRU, 314 = layer/pool). Expected sub-second epoch time for this small temporal load; 65s indicates the hang is native artifact compilation, not throughput.",
  "rationale": "Performance packet: stable 65s on nv0, replay crosses slow-runtime threshold. Earliest failing operation is compile_native_artifact during compile_model through quantized pool blocks (`.iq(2).xxs`, `.qi(4).k.m`). Existing nv0 stall issues represent residual, pool/attention, pool/GRU, or layer/pool paths, all with integer arithmetic; this reproduction uses pure pool quantization with `.bf(16)`, which is not represented. The loader, seed, loss, and bfloat16 precision execute after the compilation stall and do not establish a different public defect. Measured 65s versus expected <1s for this workload and GPU; dominant gap is NVIDIA native compiler hang, not arithmetic work or memory traffic."
}
```

## Deterministic failure packet

kind=performance
id=57cb73c0d45fb976
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10087 next:10088 composition:17856690329093
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 65 seconds
measurement=elapsed_seconds:65 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-nv0-10087-349.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/daily_shards")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.pool(2).gelu().norm(batch).iq(2).xxs
		.pool(2).norm(batch).qi(4).k.m
		.loss(bce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(17856690329093)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.bf(16)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Failed on Thursday, August 27, 2026 2:35 PM

Commit: 8f386573e4b6979d7e4d03204602144888a2ebe6

Command:
`recipe test.rs`

Reproduction script:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/daily_shards")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.pool(2).gelu().norm(batch).iq(2).xxs
		.pool(2).norm(batch).qi(4).k.m
		.loss(bce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(17856690329093)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.bf(16)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```

Full output:
```text
recipe: unknown command "test.rs"; expected `probe`

Recipe command line

Usage:
  recipe probe [OPTIONS]

The zero-argument probe uses the embedded theoretical seed contract, discovers
the current bare-metal machine, benchmarks every discovered device and link,
and writes an identity-keyed measured profile under the user's private cache.

Options:
  --contract PATH       Use another theoretical probe seed contract
  --profile PATH        Write/load this absolute measured-profile path
  --cuda-driver PATH    Exact CUDA Driver library candidate (repeatable)
  --hsa-runtime PATH    Exact ROCr/HSA runtime candidate (repeatable)
  --llvm-opt PATH       Exact LLVM IR verifier
  --llvm-llc PATH       Exact LLVM code generator
  --lld PATH            Exact ELF linker
  --ptxas PATH          Exact NVIDIA PTX assembler
  -h, --help             Show this help
```

Verdict: FAIL
Reason: The required public entrypoint rejects the exact reproduction before compilation and only accepts `probe`. The quantized pool, batch-normalization, and bf16 NVIDIA path never executes, so a finite one-epoch result and the expected sub-second runtime are unproven.

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #426.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #426: Verify current NVIDIA quantized pool execution</summary>

Closes #316

This is a verification-only PR. It intentionally changes no tracked files.

## Current public result

Commit: `9098de8711552004bb9eabd84c6d8d9c120d4c58`

Command:

```text
RECIPE_DEBUG=1 ./target/release/recipe --device benji:nv0 /home/nate/.cache/recipe/issues/316/reproduction.rs
```

The exact issue model, data path, broadcast setting, training settings, and BF16 arithmetic complete successfully from cold native artifacts:

- The BF16 calibration artifact begins with `cache=miss`.
- The 2,880-row issue artifact also begins with `cache=miss`.
- Clang emits NVIDIA LLVM bitcode, and `llc` emits PTX for both artifacts.
- The training epoch completes in 22.827 ms.
- Final BCE is finite at 0.6931.
- The process exits 0 after 6.46 seconds.

The earlier audit invoked the unrelated probe-only `/usr/bin/recipe`, so it never tested this script. The correct checkout binary reaches the full public training boundary, and the reported 65-second NVIDIA native-artifact stall does not reproduce on current `minimal`. No production-code change is supported by this result.


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.
